File: TSSR.m
function [A_final, B_final, Y_pre] = TSSR(Y, W, Sd, St, lambda_d, lambda_t, beta, n_iter, repeat_times) 
%
% Inputs:
%   Y: drug-targer association matrix.
%   W: weight matrix.
%   Sd: similarity matrix of drug.
%   St: similarity matrix of target.
%   lambda_d: parameter to control the distance between A and Sd. The default value
%   is 1.
%   lambda_t: parameter to control the distance between B and St. The default value
%   is 1.
%   beta: rate parameter of the sparse regularization. The default
%   value is 1.
%   n_iter: the number of iterations limited in RSRGM. The default value is
%   150.
%   repeat_times: repeat the algorithm with different inilization.

% Outputs:
%   A_final: sparse representation of drugs.
%   B_final: sparse representation of targets.
%   Y_pre: the predicted drug-target interaction matrix.

    if nargin < 9
        repeat_times = 5;
    end

    if nargin < 8
        n_iter = 150;
    end

    if nargin < 7
        beta = 1;
    end
    
    if nargin < 6
        lambda_t = 1;
    end
    
    if nargin < 5
        lambda_d = 1;
    end
    
    if nargin < 4
        error('Yor need input training sets and similarity matrices.');
    end
    lowest_score = inf;
    
 for i = 1: repeat_times
    fprintf(['This is the ',num2str(i), '-th repeat...'])
    fprintf('\n')

    [n,m] = size(Y); %n-drug and m-target

    A_old = Sd;
    B_old = St;

    A_old = A_old./repmat(sum(A_old,2),1,n);
    B_old = B_old./repmat(sum(B_old),m,1);

    for j  = 1: n_iter
        
        Dan = 2*(W.*Y)*(B_old'*Y') + 2*lambda_d*Sd;
        Dap = 2*(W.*(A_old*Y*B_old))*B_old'*Y' + 2*lambda_d*A_old + beta;
        aa = diag(sum(A_old./(Dap + eps),2));
        ba = sum(A_old.*(Dan./(Dap + eps)),2);
        A = A_old.*((aa*Dan + 1)./(aa*Dap + repmat(ba,1,n) + eps));
        
        Dbn = 2*(Y'*A')*(W.*Y) + 2*lambda_t*St;
        Dbp = 2*Y'*A'*(W.*(A*Y*B_old)) + 2*lambda_t*B_old + beta;
        ab = diag(sum(B_old./(Dbp + eps)));
        bb = sum(B_old.*(Dbn./(Dbp + eps)));
        B = B_old.*((Dbn*ab + 1)./(Dbp*ab + repmat(bb,m,1) + eps));

        if sum(sum(abs(A - A_old))) + sum(sum(abs(B - B_old))) < 1e-6
            break;
        else
            A_old = A;
            B_old = B;

        end
    end
    
    score = (norm(W.*(Y - A_old*Y*B_old),'fro'))^2 + lambda_d*(norm(Sd - A_old,'fro'))^2 + lambda_t*(norm(St - B_old,'fro'))^2 + beta*sum(sum(A_old)) + beta*sum(sum(B_old)); 

        if score < lowest_score

            A_final = (A + A')/2;
            B_final = (B + B')/2;
            lowest_score = score;
        end   
 end
 Y_pre = A_final*Y*B_final;
end

File: TSSR_code/TSSR_code/TSSR_master/tssr.py
#!/usr/bin/python
#coding:utf-8
from numpy import *
import numpy as np
from scipy.stats import *

import os
import scipy.linalg as LA

class TSSR:

    def __init__(self, lambda_d=2**(-5), lambda_t=2**(-5), beta= 2**(2),max_iter=150):
        self.lambda_d = lambda_d  # default 100
        self.lambda_t = lambda_t
        self.beta=beta
        self.max_iter = max_iter

    def fix_model(self, W, intMat, drugMat, targetMat):
        size = intMat.shape
        Y=np.multiply(W,intMat)
        Sd=drugMat
        St=targetMat
        lowest_score=Inf
        n = size[0]
        m = size[1]
        A_old = drugMat
        B_old = targetMat
        yy = (A_old.sum(axis=1))
        yy1 = yy.reshape(n, 1)
        zz = (B_old.sum(axis=0))
        zz1 = zz.reshape(1, m)
        A_old = A_old / np.tile(yy1, (1, n))
        B_old = B_old / np.tile(zz1, (m, 1) )
        for i in xrange(self.max_iter):
            B_old = np.mat(B_old)
            A_old = np.mat(A_old)
            Y,W=np.mat(Y),np.mat(W)
            Sd,St=np.mat(Sd),np.mat(St)
            Dan=2*np.multiply(W,Y)*(B_old.T*Y.T)+2*self.lambda_d*Sd
            Dap=2*(np.multiply(W,A_old*Y*B_old))*B_old.T * Y.T+2*self.lambda_d * A_old + self.beta*np.ones([n,n])
            Dapeps=np.finfo(float).eps * ones([n,n])
            cc=np.array((A_old / (Dap +Dapeps)).sum(axis=1).T)
            aa = np.diag(cc[0])
            ba=(np.multiply(A_old,Dan/(Dap+Dapeps ))).sum(axis=1)
            A=np.multiply(A_old,aa*Dan + np.ones([n,n]))/(aa*Dap + np.tile(ba,(1,n))+Dapeps)
            Dbn=2*(Y.T * A.T)* (np.multiply(W,Y))+ 2 * self.lambda_t * St
            Dbp=2 * Y.T * A.T * (np.multiply(W,A*Y*B_old))+ 2* self.lambda_t * B_old + self.beta*np.ones([m,m])
            Dbpeps=np.finfo(float).eps * ones([m,m])
            dd=np.array((B_old/(Dbp+ Dbpeps)).sum(axis=0))
            ab=np.diag(dd[0])
            bb=(np.multiply(B_old,Dbn/(Dbp+Dbpeps))).sum(axis=0)
            B= np.multiply(B_old,(Dbn*ab +np.ones([m,m]))/(Dbp* ab + np.tile(bb,(m,1))+ Dbpeps))

            if (abs(A-A_old)).sum(axis=0).sum(axis=1)+ (abs(B-B_old)).sum(axis=0).sum(axis=1)< 1e-6:
                break
            A_old=A
            B_old=B

        score=np.linalg.norm(np.multiply(W,Y-A_old * Y * B_old),'fro')**2 + self.lambda_d * np.linalg.norm(Sd-A_old,'fro')**2 +  self.lambda_t * np.linalg.norm(St-B_old,'fro')**2
        + self.beta * (A_old.sum(axis=0).sum(axis=1))+ self.beta * (B_old.sum(axis=0).sum(axis=1))

        if score < lowest_score:
            A_final=(A+A.T)/2
            B_final=(B+B.T)/2
            lowest_score=score

        Y_pre= A_final * Y *  B_final
        self.predictR= np.array(Y_pre)


    def predict_scores(self, test_data, N):
        inx = np.array(test_data)
        return self.predictR[inx[:, 0], inx[:, 1]]


    def __str__(self):
        return "Model: TSSR,  lambda_d:%s, lambda_t:%s, beta:%s, max_iter:%s" % ( self.lambda_d, self.lambda_t,self.beta, self.max_iter)